### ISSUE TITLE
Auto-generate a struct that allows pipelines to control pass options.

### ISSUE THREAD

**Issue by MaheshRavishankar** [llvm/llvm-project#53145]
MLIR passes can define [`PassOptions`](https://mlir.llvm.org/docs/PassManagement/#declarative-pass-specification) in tablegen. These options are controllable through command line. The same control though is not available to pass pipelines that add passes using the `create*` methods. For example a pass can be defined in Tablegen as follows

```
struct FooPass : ... {
   ...
   Option<bool> option1 {..};
   Option<int> option2 {..};
}
```

The only way to control this is to have the following code in C++

```
struct FooPass : public FooPassBase<...> {
  FooPass(bool v1, int v2) {
     this->option1.assign(v1);
     this->option2.assign(v2);
  }
}
...
std::unqiue_ptr<Pass> createFooPass(bool v1, int v2) {
  return std::make_unique<FooPass>(v1, v2);
} 
```

This is a manual process and each pass needs to explicitly declare the require pass constructor methods and `create*` method to allows pass pipelines to control the options as well. Apart from a lot of boiler plate, each time a new option is added, the constructor and `create*` method signature changes.

A way around this is to have Tablegen automatically generate a `struct` that mirrors all the options specified. So for the above example you would have the following struct generated.

```
struct FooPassOptions {
  bool option1;
  int option2;
};
```

Then you could modify the pass constructors and `create*` methods to use this generated struct. 

Note: Ideally the constructor and the `create*` methods could also be auto-generated, but chatting offline about this with @joker-eph seems like thats not very straight-forward. So having the `struct` auto-generated will at least let passes define constructors and `create*` methods that allow easily manipulation of options from a pass pipeline.

**Comment by joker-eph** [comment]
Seems like we could auto-generate a bunch of things here indeed, @Mogball wanna have a look?

**Comment by MaheshRavishankar** [comment]
The PR above is an example of what would be good to auto-generate (I did this manually for now, but its mostly boiler plate).

**Comment by GMNGeoffrey** [comment]
I have a related, but perhaps distinct request. I'll post here first, but I can open another issue. I'd like to add a flag in my binary that takes a string exactly like the options for a pass and use that to add the pass to the pipeline. It looks like my options right now are to throw a call to initializeOptions into a new create*Pass method (and ignore the potential error return) or reinvent the parsing of those options, probably all the way at the top level as llvm::cl::parser. It looks like the parsing code is all private. Perhaps we could generate the code for parsing some pass options independent from setting them on a pass?

**Comment by mscuttari** [comment]
Is there any update about this? I can make the structures for my own passes, but at the moment there is no way to specify options in C++ for the core ones.

**Comment by Mogball** [comment]
@mscuttari Has been working on this. Can this be marked as closed?

**Comment by mscuttari** [comment]
Closed with 32c5578.
